runtime/src/legacymm/cpp/Memory.cpp

```diff
@@ -2192,6 +2192,41 @@ void updateHeapRef(ObjHeader** location, const ObjHeader* object) {
   }
 }
 
+template <bool Strict>
+void updateHeapRefInsideArray(const ArrayHeader* array, const ArrayHeader* destinationArray, int index,
+                                  int fromIndex, int toIndex, int count) {
+  ObjHeader* object = *ArrayAddressOfElementAt(array, fromIndex + index);
+  ObjHeader* old = *ArrayAddressOfElementAt(destinationArray, toIndex + index);
+  // Add extra heap ref for copied elements.
+  bool addingHeapRefCondition = fromIndex >= toIndex ? fromIndex + index >= toIndex + count : fromIndex + index < toIndex;
+  bool releaseHeapRefCondition = fromIndex >= toIndex ? toIndex + index < fromIndex : toIndex + index >= fromIndex + count;
+  if (addingHeapRefCondition && object != nullptr) {
+    addHeapRef(object);
+  }
+  *const_cast<const ObjHeader**>(ArrayAddressOfElementAt(destinationArray, toIndex + index)) = object;
+  // Release rewritten elements.
+  if (releaseHeapRefCondition && reinterpret_cast<uintptr_t>(old) > 1) {
+    releaseHeapRef<Strict>(old);
+  }
+}
+
+template <bool Strict>
+void updateHeapRefsInsideOneArray(const ArrayHeader* array, const ArrayHeader* destinationArray,
```

Comment:
Why do we need both array and destinationArray? I mean it is the same array, isn't it?

Reply:
True